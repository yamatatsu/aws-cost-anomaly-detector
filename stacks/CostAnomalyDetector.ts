import * as cdk from "@aws-cdk/core";
import * as chatbot from "@aws-cdk/aws-chatbot";
import * as sns from "@aws-cdk/aws-sns";
import * as ce from "@aws-cdk/aws-ce";

type Props = cdk.StackProps & {
  slackWorkspaceId: string;
  slackChannelId: string;
};

export class CostAnomalyDetector extends cdk.Stack {
  constructor(parent: cdk.App, id: string, props: Props) {
    super(parent, id, props);

    const { slackWorkspaceId, slackChannelId } = props;

    const topic = new sns.Topic(this, "topic");

    /**
     * It is requires to some setup to be done in the AWS Chatbot console.
     * Look https://docs.aws.amazon.com/ja_jp/AWSCloudFormation/latest/UserGuide/aws-resource-chatbot-slackchannelconfiguration.html
     * 1. Select client(Slack) and click button
     * 2. Click Allow button in slack oauth page
     * 3. You can see Workspace Id
     */
    new chatbot.SlackChannelConfiguration(this, "chatbot", {
      slackChannelConfigurationName: "CostAnomalyDetector",
      slackWorkspaceId,
      slackChannelId,
      notificationTopics: [topic],
    });

    const anomalyMonitor = new ce.CfnAnomalyMonitor(this, "AnomalyMonitor", {
      monitorName: "MyAnomalyMonitor",
      monitorType: "DIMENSIONAL",
      monitorDimension: "SERVICE",
    });

    new ce.CfnAnomalySubscription(this, "AnomalySubscription", {
      subscriptionName: "MyAnomalySubscription",
      frequency: "IMMEDIATE",
      monitorArnList: [anomalyMonitor.ref],
      subscribers: [{ type: "SNS", address: topic.topicArn }],
      // 実験的に1を設定してみる
      threshold: 1,
    });
  }
}
