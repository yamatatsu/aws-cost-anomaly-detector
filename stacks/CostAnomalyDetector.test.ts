import "@aws-cdk/assert/jest";
import * as cdk from "@aws-cdk/core";
import { SynthUtils } from "@aws-cdk/assert";
import { CostAnomalyDetector } from "./CostAnomalyDetector";

test("Snapshot Test", () => {
  const app = new cdk.App();
  const target = new CostAnomalyDetector(app, "Target", {
    slackWorkspaceId: "test-slackWorkspaceId",
    slackChannelId: "test-slackChannelId",
  });

  expect(SynthUtils.toCloudFormation(target)).toMatchSnapshot();
});
