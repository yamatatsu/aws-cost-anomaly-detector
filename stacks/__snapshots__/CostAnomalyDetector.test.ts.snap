// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Test 1`] = `
Object {
  "Resources": Object {
    "AnomalyMonitor": Object {
      "Properties": Object {
        "MonitorDimension": "SERVICE",
        "MonitorName": "MyAnomalyMonitor",
        "MonitorType": "DIMENSIONAL",
      },
      "Type": "AWS::CE::AnomalyMonitor",
    },
    "AnomalySubscription": Object {
      "Properties": Object {
        "Frequency": "IMMEDIATE",
        "MonitorArnList": Array [
          Object {
            "Ref": "AnomalyMonitor",
          },
        ],
        "Subscribers": Array [
          Object {
            "Address": Object {
              "Ref": "topic69831491",
            },
            "Type": "SNS",
          },
        ],
        "SubscriptionName": "MyAnomalySubscription",
        "Threshold": 1,
      },
      "Type": "AWS::CE::AnomalySubscription",
    },
    "chatbot8E0F2E94": Object {
      "Properties": Object {
        "ConfigurationName": "CostAnomalyDetector",
        "IamRoleArn": Object {
          "Fn::GetAtt": Array [
            "chatbotConfigurationRole13839A29",
            "Arn",
          ],
        },
        "SlackChannelId": "test-slackChannelId",
        "SlackWorkspaceId": "test-slackWorkspaceId",
        "SnsTopicArns": Array [
          Object {
            "Ref": "topic69831491",
          },
        ],
      },
      "Type": "AWS::Chatbot::SlackChannelConfiguration",
    },
    "chatbotConfigurationRole13839A29": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "chatbot.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "topic69831491": Object {
      "Type": "AWS::SNS::Topic",
    },
  },
}
`;
